services/optimization.js

```diff
@@ -0,0 +1,162 @@
+    steps: v.boolean,
+    source: v.oneOf('any', 'first'),
+    destination: v.oneOf('any', 'last'),
+    distributions: v.arrayOf(
+      v.shape({
+        pickup: v.number,
+        dropoff: v.number
+      })
+    )
+  })(config);
+
+  config.profile = config.profile || 'driving';
+
+  var path = {
+    coordinates: [],
+    approach: [],
+    bearing: [],
+    radius: [],
+    distributions: []
+  };
+
+  var waypointCount = config.waypoints.length;
+  if (waypointCount < 2 || waypointCount > 12) {
+    // updated to reflect opt restrictions
+    throw new Error(
+      'waypoints must include between 2 and 12 OptimizationWaypoints'
+    );
+  }
+
+  /**
+   * @typedef {Object} OptimizationWaypoint
+   * @property {Coordinates} coordinates
+   * @property {'unrestricted'|'curb'} [approach="unrestricted"] - Used to indicate how requested routes consider from which side of the road to approach the waypoint.
+   * @property {[number, number]} [bearing] - Used to filter the road segment the waypoint will be placed on by direction and dictates the angle of approach.
+   *   This option should always be used in conjunction with a `radius`. The first value is an angle clockwise from true north between 0 and 360,
+   *   and the second is the range of degrees the angle can deviate by.
+   * @property {number|'unlimited'} [radius] - Maximum distance in meters that the coordinate is allowed to move when snapped to a nearby road segment.
+   */
+  config.waypoints.forEach(function(waypoint) {
+    path.coordinates.push(
+      waypoint.coordinates[0] + ',' + waypoint.coordinates[1]
+    );
+
+    // join props which come in pairs
+    ['bearing'].forEach(function(prop) {
+      if (waypoint.hasOwnProperty(prop) && waypoint[prop] != null) {
+        waypoint[prop] = waypoint[prop].join(',');
+      }
+    });
+
+    ['approach', 'bearing', 'radius'].forEach(function(prop) {
+      if (waypoint.hasOwnProperty(prop) && waypoint[prop] != null) {
+        path[prop].push(waypoint[prop]);
+      } else {
+        path[prop].push('');
+      }
+    });
+  });
+
+  ['approach', 'bearing', 'radius'].forEach(function(prop) {
+    // avoid sending params which are all `;`
+    if (
+      path[prop].every(function(char) {
+        return char === '';
+      })
+    ) {
+      delete path[prop];
+    } else {
+      path[prop] = path[prop].join(';');
+    }
+  });
+
+  if (config.distributions) {
+    var distributionsList = config.distributions.map(function(dist) {
+      return [dist.pickup, dist.dropoff];
+    });
+    path.distributions.push(distributionsList.join(';'));
+  }
+
+  var query = stringifyBooleans({
+    geometries: config.geometries,
+    language: config.language,
+    overview: config.overview,
+    roundtrip: config.roundtrip,
+    steps: config.steps,
+    source: config.source,
+    destination: config.destination,
+    distributions: path.distributions.length
+      ? path.distributions[0]
+      : config.distributions,
```

Comment:
Why would we fall back to `config.distributions`?

Reply:
Checking for `config.distributions` while doing the `.join(';')` dance above makes this unnecessary, so I removed it!